src/front/spv/mod.rs

```diff
@@ -590,9 +630,74 @@ impl<I: Iterator<Item = u32>> Parser<I> {
         Ok(())
     }
 
+    /// Gets an expression handle potentially performing a load/store to perform
+    /// a scope transfer.
+    #[allow(clippy::too_many_arguments)]
+    fn get_expr_handle(
+        &self,
+        id: spirv::Word,
+        lookup: &LookupExpression,
+        block_ctx: &mut BlockContext,
+        emitter: &mut super::Emitter,
+        block: &mut crate::Block,
+        expressions: &mut Arena<crate::Expression>,
+        local_arena: &mut Arena<crate::LocalVariable>,
+        body_idx: usize,
+    ) -> Handle<crate::Expression> {
+        let expr_body_idx = block_ctx.info.get(&lookup.block_id).copied().unwrap_or(0);
+
+        // Don't need to do a load/store if the expression is in the main body
+        // or if the expression is in the same body as where the query was
+        // requested. The body_idx might actually not be the final one if a loop
+        // or conditional occurs but in those cases we know that the new body
+        // will be a subscope of the body that was passed so we can still reuse
+        // the handle and not issue a load/store.
+        if expr_body_idx == 0 || expr_body_idx == body_idx {
```

Comment:
This check isn't wrong, but it does result in a lot more spills than needed. `body_idx` could be nested in a child of `expr_body_idx`, in which case `lookup.handle` is still in scope, even if `expr_body_idx` isn't the root. 

It also generates duplicate `LocalVariable`s, one for every time an expr is referenced. Spilling an expression once should be enough.

For example:

```
;; Can we force the new SPIR-V front end to perform an unnecessary spill?

                OpCapability Shader
                OpCapability Linkage
                OpMemoryModel Logical GLSL450

         %int = OpTypeInt 32 1
        %bool = OpTypeBool
 %either_type = OpTypeFunction %int %int %int %bool %bool
     %const_1 = OpConstant %int 1

      %either = OpFunction %int None %either_type
        %left = OpFunctionParameter %int
       %right = OpFunctionParameter %int
      %which1 = OpFunctionParameter %bool
      %which2 = OpFunctionParameter %bool

    %ob_label = OpLabel
                OpSelectionMerge %merge1 None
                OpBranchConditional %which1 %then %merge1

        %then = OpLabel
         %sum = OpIAdd %int %left %right
                OpSelectionMerge %merge2 None
                OpBranchConditional %which2 %take_left %take_right

   %take_left = OpLabel
   %left_plus = OpIAdd %int %left %sum
                OpBranch %merge2

  %take_right = OpLabel
  %right_plus = OpIAdd %int %right %sum
                OpBranch %merge2

      %merge2 = OpLabel
  %selection2 = OpPhi %int %left_plus %take_left %right_plus %take_right
                OpBranch %merge1

      %merge1 = OpLabel
   %selection = OpPhi %int %const_1 %ob_label %selection2 %merge2
                OpReturnValue %selection
                OpFunctionEnd
```

This generates WGSL with spills to two generated variables `local` and `local1`, both of which are unnecessary because `_e9` (generated for `%sum`) is still in scope.

```
fn function(param: i32, param1: i32, param2: bool, param3: bool) -> i32 {
    var local: i32;
    var local1: i32;
    var phi_19_: i32;
    var phi_20_: i32;

    phi_20_ = 1;
    if (param2) {
        let _e9: i32 = (param + param1);
        local = _e9;
        local1 = _e9;
        if (param3) {
            let _e11: i32 = local;
            phi_19_ = (param + _e11);
        } else {
            let _e14: i32 = local1;
            phi_19_ = (param1 + _e14);
        }
        let _e17: i32 = phi_19_;
        phi_20_ = _e17;
    }
    let _e19: i32 = phi_20_;
    return _e19;
}
```

Reply:
If you tracked each Body's parent, then you could walk up to the root from `body_idx` and see if you found `expr_body_idx` anywhere. Or, you could have a vector in `block_ctx` of the bodies we're currently inside, and search that.

But, this isn't a bug - the code is just inefficient, not wrong. What we have now doesn't even run. So this doesn't need to be fixed to merge this PR.

Reply:
This could definitely be improved, problem is that this check is run for each time an expression is used so we need an efficient to do this, we should also figure out a way to reuse the generated local but that's even harder.